src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1610 @@
+
+  // Slow helper that converts a Location to a LocationIndex. This should be
+  // avoided. TODO remove the remaining uses of this.
+  bool updateContents(const Location& location,
+                      const PossibleContents& newContents) {
+    return updateContents(getIndex(location), newContents);
+  }
+
+  // Flow contents from a location where a change occurred. This sends the new
+  // contents to all the targets of this location, and handles special cases of
+  // flow.
+  void flowAfterUpdate(LocationIndex locationIndex);
+
+  // Add a new connection while the flow is happening. If the link already
+  // exists it is not added.
+  void connectDuringFlow(Location from, Location to);
+
+  // New links added during the flow are added on the side to avoid aliasing
+  // during the flow, which iterates on links.
+  std::vector<IndexLink> newLinks;
+
+  // New links are processed when we are not iterating on any links, at a safe
+  // time.
+  void updateNewLinks();
+
+  // Contents sent to a global location can be filtered in a special way during
+  // the flow, which is handled in this helper.
+  void filterGlobalContents(PossibleContents& contents,
+                            const GlobalLocation& globalLoc);
+
+  // Reads from GC data: a struct.get or array.get. This is given the type of
+  // the read operation, the field that is read on that type, the known contents
+  // in the reference the read receives, and the read instruction itself. We
+  // compute where we need to read from based on the type and the ref contents
+  // and get that data, adding new links in the graph as needed.
+  void readFromData(HeapType declaredHeapType,
+                    Index fieldIndex,
+                    const PossibleContents& refContents,
+                    Expression* read);
+
+  // Similar to readFromData, but does a write for a struct.set or array.set.
+  void writeToData(Expression* ref, Expression* value, Index fieldIndex);
+
+  // Special handling for RefCast during the flow: RefCast only admits valid
+  // values to flow through it.
+  void flowRefCast(const PossibleContents& contents, RefCast* cast);
+
+  // We will need subtypes during the flow, so compute them once ahead of time.
+  std::unique_ptr<SubTypes> subTypes;
+
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+  // Dump out a location for debug purposes.
+  void dump(Location location);
+#endif
+};
+
+Flower::Flower(Module& wasm) : wasm(wasm) {
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "parallel phase\n";
+#endif
+
+  // First, collect information from each function.
+  ModuleUtils::ParallelFunctionAnalysis<CollectedInfo> analysis(
+    wasm, [&](Function* func, CollectedInfo& info) {
+      InfoCollector finder(info);
+
+      if (func->imported()) {
+        // Imports return unknown values.
+        for (Index i = 0; i < func->getResults().size(); i++) {
+          finder.addRoot(ResultLocation{func, i}, PossibleContents::many());
```

Comment:
If an imported function returns a value of a type that doesn't have subtypes, such as plain old types like `i32`, can't we create `ExactType` for this?

Reply:
In principle we could, yes. There is a comment in `possible-contents.h` about it, in `combine()` I believe. While we could create `ExactType{i32}`, that doesn't add any information since without subtyping we always know the exact type. And there would be a downside to doing so: "Many" is a clear signal that we can't do any better, that we've hit the worst case. "Many" indicates that for all types. If instead we emitted `ExactType{i32}` we'd need to check more carefully if it is the worst case, by seeing it is an ExactType of a type with no subtypes - not complicated, but it's simpler the other way I think.

I can add a comment here if the other isn't enough?

Reply:
I saw your new comments. Let's talk more there.